### PR TITLE
task/1259_part7_1_new_parser_for_v2_entities_options_keyValues

### PR DESCRIPTION
**Pull request by kzangeli**
New parse method for entities, using URI param options=keyValues

### CODE REVIEW

File: test/functionalTest/cases/1259_new_json_for_v2/post_v2_entities_with_keyvalues.test [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+#         "type": "MT1
+#       }
+#     }
+#   }
+# }
+#
+# In keyValues mode, there are no metadata for attributes. If you need that, use the "normalized" method.
+# 
+#
+# An example just for Curiosity ...
```
Comment (fgalan) [inline test/functionalTest/cases/1259_new_json_for_v2/post_v2_entities_with_keyvalues.test]: Curiosity -> curiosity ?

File: test/functionalTest/cases/1259_new_json_for_v2/post_v2_entities_with_keyvalues.test [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+--SHELL-INIT--
+dbInit CB
+brokerStart CB 0-255
+
+--SHELL--
+
+#
+# We now have two different ways of parsing "POST /v2/entities":
+#
+# 1. The 'old way', or "Normalized", which is the default parsing method:
```
Comment (fgalan) [inline test/functionalTest/cases/1259_new_json_for_v2/post_v2_entities_with_keyvalues.test]: The SHELL preamble comments are a very good piece of documentation from a "pragmatic usage of the API" point of view. Maybe the text should be reused in some part of the documentation, e.g. the NGSIv2 cookbook.

File: src/lib/jsonParseV2/parseContextAttribute.cpp [hunk 3]
```diff
@@ -121,11 +121,54 @@ static std::string parseContextAttributeObject(const Value& start, ContextAttrib
         return r;
       }
     }
-    else  // ERROR
+    else if (keyValues == false) // ERROR
     {
       LM_W(("Bad Input (unrecognized property for ContextAttribute - '%s')", name.c_str()));
       return "unrecognized property for context attribute";
     }
+    else  // Anything captured here is a metadata, and options=keyValues has been set
```
Comment (fgalan) [inline src/lib/jsonParseV2/parseContextAttribute.cpp]: Not sure of understand the comment "Anything captured here is a metadata, and options=keyValues has been set" as is keyValues is set, then no metadata should be captured... maybe I'm missing something. Could you elaborate, please?

File: src/lib/jsonParseV2/parseEntity.cpp [hunk 4]
```diff
@@ -49,6 +49,10 @@ using namespace rapidjson;
 *
 * In the case of /v2/entities/<eid>, the entityId of 'Entity* eP' is set in
 * the service routine postEntity.
+*
+* Also, if the URI param 'options' includes the value 'keyValues', then the
```
Comment (fgalan) [inline src/lib/jsonParseV2/parseEntity.cpp]: The part "or metadata inside attribute" is not correct, as keyValues only allows to "compact" attribute values, but not metadata. Or maybe I'm not understanding correctly this comment.

File: test/functionalTest/cases/1187_empty_attribute_in_create/empty_attribute_in_create.test [hunk 5]
```diff
@@ -120,11 +120,6 @@ Date: REGEX(.*)
         "type": null,
         "value": ""
     },
-    "A2": {
```
Comment (fgalan) [inline test/functionalTest/cases/1187_empty_attribute_in_create/empty_attribute_in_create.test]: I understand that the "side effect" which causes A2 not being shown, should be fixed.

File: src/lib/jsonParseV2/parseContextAttribute.cpp (hunk 3 above)
Comment (kzangeli) [inline src/lib/jsonParseV2/parseContextAttribute.cpp]: We are inside an attribute here (and keyValues is *true*), so if a field named "m1" is encountered, then it is not an error, but a metadata (without keyValues true) it would be an error.

File: src/lib/jsonParseV2/parseEntity.cpp (hunk 4 above)
Comment (kzangeli) [inline src/lib/jsonParseV2/parseEntity.cpp]: Maybe I'm the one that has misunderstood. 
I am treating 'metadatas inside attributes' just like 'attributes inside entities'.

File: test/functionalTest/cases/1259_new_json_for_v2/post_v2_entities_with_keyvalues.test (hunk 1 above)
Comment (kzangeli) [inline test/functionalTest/cases/1259_new_json_for_v2/post_v2_entities_with_keyvalues.test]: sure :-)

File: test/functionalTest/cases/1259_new_json_for_v2/post_v2_entities_with_keyvalues.test (hunk 2 above)
Comment (kzangeli) [inline test/functionalTest/cases/1259_new_json_for_v2/post_v2_entities_with_keyvalues.test]: With some minor changes this chunk of text could be part of out documentation, yes.  
Tell me where exactly and I'll put it in.

File: test/functionalTest/cases/1187_empty_attribute_in_create/empty_attribute_in_create.test (hunk 5 above)
Comment (kzangeli) [inline test/functionalTest/cases/1187_empty_attribute_in_create/empty_attribute_in_create.test]: Actually, there has been a bug fix.
We talked about this ...
Before, the broker signaled error and created A2 (in the previous step).
Now, the same error is signaled but A2 is not created.

This is better than before.

Even better would be to NOT signal error and to create A2, but that is a bigger fix and not included in this PR.

Comment (fgalan) [inline test/functionalTest/cases/1187_empty_attribute_in_create/empty_attribute_in_create.test]: You are right... Actually, the fix is correct.

Maybe it would be clerarer chaning a little bit the step descriptions. In particular:

03. POST /v2/entities to create E1/A2="", using json object

to

03. POST /v2/entities to create E1/A2="", using json object (fails, as entity already exists)

And:

04. GET /v2/entities/E1 to see A2==""

to

04. GET /v2/entities/E1 to not see A2

Comment (kzangeli) [inline test/functionalTest/cases/1187_empty_attribute_in_create/empty_attribute_in_create.test]: Of course. The 'descriptive texts' haven't been modified. just a lapsus on my behalf.

File: src/lib/jsonParseV2/parseContextAttribute.cpp (hunk 3 above)
Comment (kzangeli) [inline src/lib/jsonParseV2/parseContextAttribute.cpp]: After a short discussion on skype I have realized my mistake.
I still think 'the old way' I'm afraid ...
Will be fixed.

File: src/lib/jsonParseV2/parseEntity.cpp (hunk 4 above)
Comment (kzangeli) [inline src/lib/jsonParseV2/parseEntity.cpp]: Yes, I am definitely the one that has misunderstood. To be fixed.

File: test/functionalTest/cases/1187_empty_attribute_in_create/empty_attribute_in_create.test (hunk 5 above)
Comment (kzangeli) [inline test/functionalTest/cases/1187_empty_attribute_in_create/empty_attribute_in_create.test]: Fixed in 47a8a42

File: src/lib/jsonParseV2/parseEntity.cpp (hunk 4 above)
Comment (kzangeli) [inline src/lib/jsonParseV2/parseEntity.cpp]: Fixed in 47a8a42

File: src/lib/jsonParseV2/parseContextAttribute.cpp (hunk 3 above)
Comment (kzangeli) [inline src/lib/jsonParseV2/parseContextAttribute.cpp]: Fixed in 47a8a42

File: test/functionalTest/cases/1259_new_json_for_v2/post_v2_entities_with_keyvalues.test (hunk 1 above)
Comment (kzangeli) [inline test/functionalTest/cases/1259_new_json_for_v2/post_v2_entities_with_keyvalues.test]: Fixed in 47a8a42

File: test/functionalTest/cases/1259_new_json_for_v2/post_v2_entities_with_keyvalues.test (hunk 2 above)
Comment (fgalan) [inline test/functionalTest/cases/1259_new_json_for_v2/post_v2_entities_with_keyvalues.test]: In order not to block this PR, I'd suggest to do it in a separate PR.

File: src/lib/jsonParseV2/parseEntity.cpp [hunk 6]
```diff
@@ -49,6 +49,9 @@ using namespace rapidjson;
...
+* parse changes to accept 'unknown fields' as attributes inside entities.
```
Comment (fgalan) [inline src/lib/jsonParseV2/parseEntity.cpp]: "Also, if the URI param 'options' includes the value 'keyValues', then the parse changes to accept 'unknown fields' as attributes inside entities".

Not sure if the comment is fully correct... I mean, in no-keyValues mode, the attributes are also unkown fields (the difference is that the value of such unknown fields follows a given pattern). Maybe I don't understanding correctly...

Comment (kzangeli) [inline src/lib/jsonParseV2/parseEntity.cpp]: True.
I guess I was thinking about the way I wanted to implement the metadatas when I wrote this.

File: test/functionalTest/cases/0979_patch_v2_entities_eid/patch_v2_entities_eid.test [hunk 7]
```diff
@@ -35,7 +35,8 @@ brokerStart CB
 # 03. GET /v2/entities/E1
 # 04. PATCH /v2/entities/E1 attr1=4
 # 05. GET /v2/entities/E1 - see attr1=4 AND attr2=2 (untouched)
-# 06. PATCH /v2/entities/E1 attr3=6 - error, as attr3 doesn't exist
+# 06. PATCH /v2/entities/E1 attr3=6
```
Comment (fgalan) [inline test/functionalTest/cases/0979_patch_v2_entities_eid/patch_v2_entities_eid.test]: I think this should be rolled-back... PATCH cannot be used to add new attributes, so if attr3 doesn't previously exist, then it should lead to an error.

File: test/functionalTest/cases/0979_patch_v2_entities_eid/patch_v2_entities_eid.test [hunk 8]
```diff
@@ -35,7 +35,8 @@ brokerStart CB
...
+# 07. GET /v2/entities/E1 - see attr1, attr2 and attr3
```
Comment (fgalan) [inline test/functionalTest/cases/0979_patch_v2_entities_eid/patch_v2_entities_eid.test]: attr3 will not be shown, after rollback (probably the step 07 could be removed, to revert this .test to its original form).

File: test/functionalTest/cases/1187_empty_attribute_in_create/empty_attribute_in_create.test [hunk 9]
```diff
@@ -32,8 +32,8 @@ brokerStart CB
 #
 # 01. POST /v2/entities to create E1/A1=""
 # 02. GET /v2/entities/E1 to see A1==""
-# 03. POST /v2/entities to create E1/A2="", using json object
-# 04. 04. GET /v2/entities/E1 to see A2==""
+# 03. POST /v2/entities to create E1/A2="", using json object (fails, as entity already exists)
```
Comment (fgalan) [inline test/functionalTest/cases/1187_empty_attribute_in_create/empty_attribute_in_create.test]: This should be roll-backed... According to specification, POST on /v2/entities/E1 will lead to append/update attributes. Similarly, step 03 should be also rollbacked. Probably, leaving this .test in its original form will work.

File: test/functionalTest/cases/1259_new_json_for_v2/post_v2_entities_eid_with_keyvalues.test [hunk 10]
```diff
@@ -0,0 +1,272 @@
...
+dbInit CB
+brokerStart CB 0-255
+
+--SHELL--
+
+#
+# 01. POST /v2/entities, creating E1
+# 02. PUT /v2/entities/E1?options=keyValues, adding A1 to E1
+# 03. GET /v2/entities/E1
+# 04. PATCH /v2/entities/E1?options=keyValues, adding A2 to E1
```
Comment (fgalan) [inline test/functionalTest/cases/1259_new_json_for_v2/post_v2_entities_eid_with_keyvalues.test]: It should be POST (PATCH is not for adding)

File: test/functionalTest/cases/1259_new_json_for_v2/post_v2_entities_eid_with_keyvalues.test [hunk 11]
```diff
@@ -0,0 +1,272 @@
...
+
+--SHELL-INIT--
+dbInit CB
+brokerStart CB 0-255
+
+--SHELL--
+
+#
+# 01. POST /v2/entities, creating E1
+# 02. PUT /v2/entities/E1?options=keyValues, adding A1 to E1
```
Comment (fgalan) [inline test/functionalTest/cases/1259_new_json_for_v2/post_v2_entities_eid_with_keyvalues.test]: Maybe this PUT could be changed to PATCH (PUT will also work, but it is properly tested in "replace" semantics in step 6).

File: test/functionalTest/cases/1260_v2_update_patch_error_description/update_patch_error_description.test [hunk 12]
```diff
@@ -88,15 +88,10 @@ Date: REGEX(.*)
 
 03. PATCH /v2/entities/Eno (not found)
```
Comment (fgalan) [inline test/functionalTest/cases/1260_v2_update_patch_error_description/update_patch_error_description.test]: I think this test should be rollbacked.

File: test/functionalTest/cases/1278_patch_unknown_attribute/patch_unknown_attribute.test [hunk 13]
```diff
@@ -31,8 +31,9 @@ brokerStart CB
 
 #
 # 01. POST /v2/entities to create E1/T1/A1=0
-# 02. PATCH /v2/entities/E1 (A2 unknown)
-
+# 02. PATCH /v2/entities/E1 (to create attr A2)
```
Comment (fgalan) [inline test/functionalTest/cases/1278_patch_unknown_attribute/patch_unknown_attribute.test]: I think this .test should be rollbacked.

File: test/functionalTest/cases/1187_empty_attribute_in_create/empty_attribute_in_create.test (hunk 9 above)
Comment (kzangeli) [inline test/functionalTest/cases/1187_empty_attribute_in_create/empty_attribute_in_create.test]: Except there is no POST on /v2/entities/E1.
There is a second POST on /v2/entities (with E1 in payload).

I understand a second POST /v2/entities with the same *id* in the payload should give an error ...

Comment (kzangeli) [inline test/functionalTest/cases/1187_empty_attribute_in_create/empty_attribute_in_create.test]: Discussed on skype. All OK

NTC

File: src/lib/jsonParseV2/parseEntity.cpp (hunk 6 above)
Comment (kzangeli) [inline src/lib/jsonParseV2/parseEntity.cpp]: Fixed in a5f5402

File: test/functionalTest/cases/0979_patch_v2_entities_eid/patch_v2_entities_eid.test (hunk 7 above)
Comment (kzangeli) [inline test/functionalTest/cases/0979_patch_v2_entities_eid/patch_v2_entities_eid.test]: Fixed in a5f5402

File: test/functionalTest/cases/0979_patch_v2_entities_eid/patch_v2_entities_eid.test (hunk 8 above)
Comment (kzangeli) [inline test/functionalTest/cases/0979_patch_v2_entities_eid/patch_v2_entities_eid.test]: Fixed in a5f5402

File: test/functionalTest/cases/1259_new_json_for_v2/post_v2_entities_eid_with_keyvalues.test (hunk 10 above)
Comment (kzangeli) [inline test/functionalTest/cases/1259_new_json_for_v2/post_v2_entities_eid_with_keyvalues.test]: Fixed in a5f5402

File: test/functionalTest/cases/1259_new_json_for_v2/post_v2_entities_eid_with_keyvalues.test (hunk 11 above)
Comment (kzangeli) [inline test/functionalTest/cases/1259_new_json_for_v2/post_v2_entities_eid_with_keyvalues.test]: Fixed in a5f5402

File: test/functionalTest/cases/1260_v2_update_patch_error_description/update_patch_error_description.test (hunk 12 above)
Comment (kzangeli) [inline test/functionalTest/cases/1260_v2_update_patch_error_description/update_patch_error_description.test]: Fixed in a5f5402

File: test/functionalTest/cases/1278_patch_unknown_attribute/patch_unknown_attribute.test (hunk 13 above)
Comment (kzangeli) [inline test/functionalTest/cases/1278_patch_unknown_attribute/patch_unknown_attribute.test]: Fixed in a5f5402

File: src/lib/mongoBackend/MongoCommonUpdate.cpp [hunk 14]
```diff
@@ -2921,6 +2921,11 @@ void processContextElement
         buildGeneralErrorResponse(ceP, NULL, responseP, SccInvalidModification, "Already Exists");
         return;
     }
+    else if ((checkEntityExistance == true) && (action == "APPEND") && (entitiesNumber == 0))
```
Comment (fgalan) [inline src/lib/mongoBackend/MongoCommonUpdate.cpp]: Why don't just `checkEntityExistance` instead of `checkEntityExistance == true`? ;)

File: test/functionalTest/cases/1259_new_json_for_v2/post_v2_entities_eid_with_keyvalues.test [hunk 15]
```diff
@@ -0,0 +1,274 @@
...
+POST /v2/entities with key-values
+
+--SHELL-INIT--
+dbInit CB
+brokerStart CB 0-255
+
+--SHELL--
+
+#
+# 01. POST /v2/entities, creating E1
```
Comment (fgalan) [inline test/functionalTest/cases/1259_new_json_for_v2/post_v2_entities_eid_with_keyvalues.test]: Maybe PATCH testing should be added to this sequence?

File: src/lib/mongoBackend/MongoCommonUpdate.cpp (hunk 14 above)
Comment (kzangeli) [inline src/lib/mongoBackend/MongoCommonUpdate.cpp]: Matter of opinion. I prefer to have the '== true' there.
If you want to, we decide a way to do this and we add it to the style guide.
I don't have an especially strong opinion.
I prefer adding the == true for clarity but I know most people dislike it.

File: test/functionalTest/cases/1259_new_json_for_v2/post_v2_entities_eid_with_keyvalues.test (hunk 15 above)
Comment (kzangeli) [inline test/functionalTest/cases/1259_new_json_for_v2/post_v2_entities_eid_with_keyvalues.test]: Fixed in 2d3d4cd

**Comment by fgalan** [comment]
LGTM
